roles/forklift_versions/lookup_plugins/forklift_versions.py

```diff
@@ -0,0 +1,94 @@
+          scenario_version:
+            description: scenario version to look up
+        notes:
+          - this lookup will match the foreman version for the foreman scenatio
+          - this lookup will match the katello version for all other scenarios
+"""
+import yaml
+from ansible.errors import AnsibleParserError, AnsibleLookupError
+from ansible.plugins.lookup import LookupBase
+
+try:
+    from __main__ import display
+except ImportError:
+    from ansible.utils.display import Display
+    display = Display()
+
+
+def version_sort_key(version):
+    try:
+        return [int(u) for u in version.split('.')]
+    except ValueError:
+        return [9999, 9999, 9999]
+
+
+class LookupModule(LookupBase):
+
+    def run(self, terms, variables=None, **kwargs):
+        ret = []
+
+        for term in terms:
+            display.debug("Forklift lookup term: %s" % term)
+
+            lookup_params = dict(x.split('=', 1) for x in term.split())
+
+            try:
+                scenario = lookup_params['scenario']
+                if scenario != 'foreman':
+                    scenario = 'katello'
+                scenario_version = lookup_params['scenario_version']
+                versions_file_name = lookup_params['file']
+                upgrade = lookup_params.get('upgrade', False)
+            except KeyError:
+                raise AnsibleParserError("missing required param for forklift_version")
+
+            try:
+                with open(versions_file_name, 'r') as versions_file:
+                    versions = yaml.safe_load(versions_file)
+            except Exception:
+                raise AnsibleLookupError("couldn't read '%s'" % versions_file_name)
+
+            if not upgrade:
+                for version in versions['installers']:
+                    if version[scenario] == scenario_version:
+                        forklift_vars = {
+                                'foreman_repositories_version': version['foreman'],
+                                'foreman_client_repositories_version': version['foreman'],
+                                'katello_repositories_version': version['katello'],
+                                'katello_repositories_pulp_version': version['pulp'],
+                                'pulp_repositories_version': version['pulp'],
+                                'puppet_repositories_version': version['puppet'],
+                                }
+                        ret.append(forklift_vars)
+                        break
+            else:
+                upgrade_versions = set()
+                for version in reversed(versions['installers']):
+                    if version[scenario] == scenario_version:
+                        upgrade_versions.add(scenario_version)
+                    elif 1 <= len(upgrade_versions) < 3:
+                        upgrade_versions.add(version[scenario])
+                upgrade_versions = sorted(upgrade_versions, key=version_sort_key)
```

Comment:
Why do we sort the versions in the end? Shouldn't we sort installers since the loop already relies on order.

Reply:
This is Evgeni being SMRT ;-)

`upgrade_versions` starts its life as a `set`, so that when I loop over the installers, I can add them as they come and ignore the fact that we have Foreman 1.20 twice. However, when you convert a `set` to a `list`, it seems to lose insertion order (or rather, a `set` is alphabetically ordered):

```
>>> list(set(['1.20', '1.20', '1.21', 'nightly']))
['nightly', '1.21', '1.20']
>>> list(set(['3.9', '3.10', '3.11']))
['3.10', '3.11', '3.9']
```

Reply:
Oh yes, there is no SortedSet. Carry on.